src/components/lib/hw_perf/arch/x86_64/hw_perf.c

```diff
@@ -0,0 +1,50 @@
+
+#include <cos_types.h>
+#include <hw_perf.h>
+#include <cos_kernel_api.h>
+
+
+/* 
+ * x86_64 hardware instruction to read out a performance counter 
+ * 
+ * cntr = 0->7		: programmable counters 0 through 7 respectivly
+ * cntr = 1<<30		: fixed counter 0 (hardware instruction counter on most intel archs)
+ * cntr = (1<<30)+1	: fixed counter 1 (core cycles counter on most intel arches) 
+ * cntr = (1<<30)+2	: fixed counter 2 (reference cycles counter on most intel arches) 
+ *
+ * Note: documentation for this is hard to find; I referenced the intel x64 manual and this thread:
+ * https://community.intel.com/t5/Software-Tuning-Performance/How-to-read-performance-counters-by-rdpmc-instruction/td-p/1009043
+ *
+ * */
+static inline unsigned long 
+rdpmc (unsigned long cntr)
+{
+	unsigned int low, high;
+
+	asm volatile("rdpmc" : "=a" (low), "=d" (high) : "c" (cntr));
+
+	return low | ((unsigned long)high) << 32;
```

Comment:
I don't want to lookup associativity rules to know if this correct every time. Can you add `()` to make sure that the `<<` happens first. I did look it up, and this is fine, but I'd prefer that we set a standard of making this type of thing explicit.